### DISCUSSION TITLE
Unexpected behaviour with 2 mrouted's

### DISCUSSION THREAD

**Discussion by lucdig** [troglobit/mrouted#59] [Q&A]
Hi, I'm testing this configuration:

generator ---> mrouted 1 <--- tunnel ---> mrouted 2 ---> receiver

A) Unexpected behaviour:
```
1 - generator sends 239.12.0.1
2 - mrouted 1 receives 239.12.0.1
3 - mrouted 1 forwards 239.12.0.1 to mrouted 2, without any join
4 - receiver joins 239.12.0.1
5 - mrouted 2 forwards 239.12.0.1 to receiver
```

B) Expected behaviour:
```
1 - generator sends 239.12.0.1
2 - mrouted 1 receives 239.12.0.1
3 - mrouted 1 does not forward 239.12.0.1 to mrouted 2 yet
4 - receiver joins 239.12.0.1 
5 - mrouted 2 joins 239.12.0.1 (to mrouted 1)
6 - mrouted 1 forwards 239.12.0.1 to mrouted 2
7 - mrouted 2 forwards 239.12.0.1 to receiver
```

I expected to have traffic between mrouted 1 and mrouted 2 only when a receiver joins a Mgroup, but I see that traffic between mrouted 1 and mrouted 2 is forwarded anyway.

Am I missing some configuration? Or, is it normal that traffic between mrouted's is forwarded anyway?

I would like mrouted 1 to forward traffic to mrouted 2 only if mrouted 2 receives an igmp join.
Is it possible?

Thanks in advance for your response, regards

**Comment by troglobit** [comment]
mrouted is an implementation of DVMRP which is a flood-and-prune protocol. So yeah, this is expected behavior.

**Comment by lucdig** [comment]
I see. I'm on OpenBSD 7.2 with your version compiled.

The flow should be:

```
1 - generator sends 239.12.0.1
2 - mrouted 1 receives 239.12.0.1
3 - mrouted 1 forwards 239.12.0.1 to mrouted 2
4 - mrouted 2 sends a Prune to mrouted 1 and 239.12.0.1 is not forwarded anymore
5 - receiver joins 239.12.0.1
6 - mrouted 2 sends a Graft to mrouted 1 for 239.12.0.1
7 - mrouted 2 receives Graft-Ack from mrouted 1
8 - mrouted 1 forwards 239.12.0.1 to mrouted 2
9 - mrouted 2 forwards 239.12.0.1 to receiver
```

Anyway, PF blocks the Prune message, even if I have, in /etc/pf.conf:

`pass proto igmp allow-opts`

In debugs I see a **Permission denied**:

```
08:53:04.212 warning - sendto to 10.11.12.2 on 10.11.12.1: Permission denied
08:53:04.213 SENT prune message      from 10.11.12.1      to 10.11.12.2
08:53:04.213 sent prune for (10.0.12/24 239.12.0.1)/400 on vif 2 to 10.11.12.2
```

If I disable packet filter (pfctl -d), Prune message is **not** blocked and everything works as expected.

Do you know if any additional rule should be configured in /etc/pf.conf, to have Prune not blocked?

Thanks again, regards

**Reply by troglobit** [reply]
No, that I do not know.  I try to avoid using firewalls/NAT and multicast routing.  You'll have to ask someone with experience in OpenBSD pf.

**Reply by lucdig** [reply]
Ok, thanks a lot anyway. I've sent an email to bugs@openbsd.org
Regards

**Reply by lucdig** [reply]
Hi, for your info, from bugs@openbsd.org I received a response regarding Prunes.
In https://www.openbsd.org/plus72.html you can find:
"Made pf(4) more paranoid about IGMP/MKP messages."
They made pf too paranoid, because it drops all IGMP messages with TTL!=1 or IP Destination Address not 224/4.
I hope that soon this bug will be fixed.

On the other hand, I see a strange behaviour in you last mrouted, installed on a Linux.
After receving Prunes from other mrouted's, it floods traffic after 5-10 minutes (don't know exactly) without any Graft received.

```
generator -- ep1 -- mrouted_linux ----+---- mrouted1
                                      |
                                      +---- mrouted2
```
mrouted_linux is mrouted version 4.5-dev
generator and mrouted_linux are on the same Linux, connected by the ep1 interface.

**mrouted.conf:**
noflood
cache-lifetime 3600

mrouted1 and mrouted2 are OpenBSD v3.8

1 - generator sends traffic 239.254.254.1
2 - mrouted_linux floods 239.254.254.1 to mrouted1 and mrouted2
3 - mrouted1 and mrouted2 send Prunes
4 - mrouted_linux stops flooding 239.254.254.1 to mrouted1 and mrouted2
5 - after 5-10 mins: mrouted_linux starts again flooding 239.254.254.1 to mrouted1 and mrouted2
6 - on mrouted1 and mrouted2 I see: "warning: kernel entry already exists", and they do not send Prunes again

So, the traffic flooded again in point 5 - is not pruned, even with noflood set in mrouted.conf of mrouted_linux.

Do you have any idea why 5 - happens?

Thanks
Luca

**Reply by troglobit** [reply]
OpenBSD based their mrouted on an older version.  "Mine" is based on Bill Fenner's v3.9-beta3 release which has many more features that were never officially incorporated in any RFC post [RFC1075](https://www.rfc-editor.org/rfc/rfc1075.html).  For instance, I don't believe the OpenBSD one has support for the `cache-lifetime` and `prune-lifetime` settings, which I believe control the behavior you are seeing.  Much of these settings are for fault-tolerance reasons, e.g. routers restarting.

**Reply by lucdig** [reply]
I see.

Actually, with 4.5 compiled in OpenBSD, I see that after 5 - there are Prunes by mrouted1 and mrouted2.

Thanks again, regards

**Reply by lucdig** [reply]
fyi, in OpenBSD 7.3 Prune, Graft and Graft-Ack messages are not blocked anymore by PF
